data_structures/reverse_a_linked_list.cpp

```diff
@@ -0,0 +1,208 @@
+ */
+void list::insert(int32_t n) {
+    try {
+        Node *new_node = new Node();
+        Node *temp = nullptr;
+        new_node->val = n;
+        new_node->next = nullptr;
+        if (isEmpty()) {
+            head = new_node;
+        } else {
+            temp = head;
+            while (temp->next != nullptr) {
+                temp = temp->next;
+            }
+            temp->next = new_node;
+        }
+    } catch (std::bad_alloc &exception) {
+        std::cerr << "bad_alloc detected: " << exception.what() << "\n";
+    }
+}
+
+/**
+ * @brief Utility function for reversing a list
+ * @brief Using the current, previous, and next pointer.
+ * @returns void
+ */
+void list::reverseList() {
+    Node *curr = head;
+    Node *prev = nullptr, *next_node = nullptr;
+    while (curr != nullptr) {
+        next_node = curr->next;
+        curr->next = prev;
+        prev = curr;
+        curr = next_node;
+    }
+    head = prev;
+}
+
+/**
+ * @brief Utility function to find the top element of the list
+ * @returns the top element of the list
+ */
+int32_t list::top() {
+    try {
+        if (!isEmpty()) {
+            return head->val;
+        }
+    } catch (const std::exception &e) {
+        std::cerr << "List is empty" << e.what() << '\n';
+    }
+}
+/**
+ *  @brief Utility function to find the last element of the list
+ *  @returns the last element of the list
+ */
+int32_t list::last() {
+    try {
+        if (!isEmpty()) {
+            Node *t = head;
+            while (t->next != nullptr) {
+                t = t->next;
+            }
+            return t->val;
+        }
+    } catch (const std::exception &e) {
+        std::cerr << "List is empty" << e.what() << '\n';
+    }
+}
+/**
+ *  @brief Utility function to find the i th element of the list
+ *  @returns the i th element of the list
+ */
+int32_t list::traverse(int index) {
+    Node *current = head;
+
+    int count = 0;
+    while (current != nullptr) {
+        if (count == index) {
+            return (current->val);
+        }
+        count++;
+        current = current->next;
+    }
+
+    /* if we get to this line,the caller was asking for a non-existent element
+    so we assert fail */
+    assert(0);
+}
+
+}  // namespace linked_list
+}  // namespace data_structures
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    data_structures::linked_list::list L;
+    // 1st test
+    L.insert(11);
+    L.insert(12);
+    L.insert(15);
+    L.insert(10);
+    L.insert(-12);
+    L.insert(-20);
+    L.insert(18);
+    assert(L.top() == 11);
+    assert(L.last() == 18);
+    L.reverseList();
+    // Reversal Testing
+    assert(L.top() == 18);
+    assert(L.traverse(1) == -20);
+    assert(L.traverse(2) == -12);
+    assert(L.traverse(3) == 10);
+    assert(L.traverse(4) == 15);
+    assert(L.traverse(5) == 12);
+    assert(L.last() == 11);
+    std::cout << "Passed" << std::endl;
```

Comment:
*Suggested previously by @amino19.*
```suggestion
    std::cout << "All tests have successfully passed!" << std::endl;
```